File: src/21.merge-two-sorted-lists.cpp
/*
 * @lc app=leetcode id=21 lang=cpp
 *
 * [21] Merge Two Sorted Lists
 */

// Definition for singly-linked list.
struct ListNode {
  int val;
  ListNode* next;
  ListNode() : val(0), next(nullptr) {}
  ListNode(int x) : val(x), next(nullptr) {}
  ListNode(int x, ListNode* next) : val(x), next(next) {}
};
// @lc code=start
class Solution {
 public:
  ListNode* mergeTwoLists(ListNode* l1, ListNode* l2) {
    ListNode* root = new ListNode();
    ListNode* current = root;

    while (l1 != nullptr && l2 != nullptr) {
      if (l1->val < l2->val) {
        current->next = l1;
        l1 = l1->next;
      } else {
        current->next = l2;
        l2 = l2->next;
      }
      current = current->next;
    }

    if (l1 == nullptr) {
      current->next = l2;
    } else {
      current->next = l1;
    }

    return root->next;
  }
};
// @lc code=end

File: Top Interview Questions/src/21.merge-two-sorted-lists.java
import jdk.jfr.internal.SecuritySupport;

/*
 * @lc app=leetcode id=21 lang=java
 *
 * [21] Merge Two Sorted Lists
 */

// Definition for singly-linked list.
class ListNode {
  int val;
  ListNode next;
  ListNode() {}
  ListNode(int val) { this.val = val; }
  ListNode(int val, ListNode next) { this.val = val; this.next = next; }
}
// @lc code=start
class Solution {
  public ListNode mergeTwoLists(ListNode l1, ListNode l2) {
    ListNode head = new ListNode();
    ListNode cursor = head;

    while (l1 != null && l2 != null) {
      if (l1.val < l2.val) {
        cursor.next = l1;
        l1 = l1.next;
      } else {
        cursor.next = l2;
        l2 = l2.next;
      }
      cursor = cursor.next;
    }

    if (l1 == null) {
      while (l2 != null) {
        cursor.next = l2;
        l2 = l2.next;
        cursor = cursor.next;
      }
    } else {
      while (l1 != null) {
        cursor.next = l1;
        l1 = l1.next;
        cursor = cursor.next;
      }
    }

    return head.next;
  }
}
// @lc code=end

